src/vs/platform/userDataProfile/common/userDataProfile.ts

```diff
@@ -0,0 +1,119 @@
+	readonly location: URI;
+	readonly globalStorageHome: URI;
+	readonly settingsResource: URI;
+	readonly keybindingsResource: URI;
+	readonly tasksResource: URI;
+	readonly snippetsHome: URI;
+	readonly extensionsResource: URI | undefined;
+}
+
+export const IUserDataProfilesService = createDecorator<IUserDataProfilesService>('IUserDataProfilesService');
+export interface IUserDataProfilesService {
+	readonly _serviceBrand: undefined;
+
+	readonly profilesHome: URI;
+	readonly defaultProfile: IUserDataProfile;
+
+	readonly onDidChangeCurrentProfile: Event<IUserDataProfile>;
+	readonly currentProfile: IUserDataProfile;
+
+	createProfile(name: string): IUserDataProfile;
+	setProfile(name: string): Promise<void>;
+	getAllProfiles(): Promise<IUserDataProfile[]>;
+
+}
+
+function reviveProfile(profile: IUserDataProfile, scheme: string): IUserDataProfile {
+	return {
+		name: profile.name,
+		location: URI.revive(profile.location).with({ scheme }),
+		globalStorageHome: URI.revive(profile.globalStorageHome).with({ scheme }),
+		settingsResource: URI.revive(profile.settingsResource).with({ scheme }),
+		keybindingsResource: URI.revive(profile.keybindingsResource).with({ scheme }),
+		tasksResource: URI.revive(profile.tasksResource).with({ scheme }),
+		snippetsHome: URI.revive(profile.snippetsHome).with({ scheme }),
+		extensionsResource: URI.revive(profile.extensionsResource)?.with({ scheme })
+	};
+}
+
+export class UserDataProfilesService extends Disposable implements IUserDataProfilesService {
+	readonly _serviceBrand: undefined;
+
+	protected static DEFAULT_PROFILE_NAME = 'default';
+
+	protected _currentProfile: IUserDataProfile;
+	get currentProfile(): IUserDataProfile { return this._currentProfile; }
+
+	readonly profilesHome: URI;
+	protected _defaultProfile: IUserDataProfile;
+	get defaultProfile(): IUserDataProfile { return this._defaultProfile; }
+
+	private readonly _onDidChangeCurrentProfile = this._register(new Emitter<IUserDataProfile>());
+	readonly onDidChangeCurrentProfile = this._onDidChangeCurrentProfile.event;
+
+	constructor(
+		defaultProfile: IUserDataProfile | undefined,
+		currentProfile: IUserDataProfile | undefined,
+		@IEnvironmentService private readonly environmentService: IEnvironmentService,
+		@IFileService protected readonly fileService: IFileService,
+		@ILogService protected readonly logService: ILogService
+	) {
+		super();
+		this.profilesHome = joinPath(this.environmentService.userRoamingDataHome, 'profiles');
+		this._defaultProfile = defaultProfile ? reviveProfile(defaultProfile, this.profilesHome.scheme) : this.createProfile(undefined);
+		this._currentProfile = currentProfile ? reviveProfile(currentProfile, this.profilesHome.scheme) : this._defaultProfile;
+	}
+
+	createProfile(name: string | undefined): IUserDataProfile {
+		const location = name && name !== UserDataProfilesService.DEFAULT_PROFILE_NAME ? joinPath(this.profilesHome, name) : this.environmentService.userRoamingDataHome;
```

Comment:
As discussed, use `hash` to get fixed length folder names for profiles and avoid issues when profile name cannot be used as folder name.

Reply:
Sure. Will consider this in my next PR.

Reply:
If you remember, fine to resolve 👍